Onboard endnode

yongpingchen opened the pull request.

yongpingchen committed:
implement onboardEndnodeWithGatewawy

yongpingchen committed:
add test

yongpingchen committed:
fix typo

satoshikumano commented on api.go:
```diff
@@ -189,6 +189,22 @@ type OnboardByOwnerRequest struct {
 }
 
+// OnboardEndnodeWithGatewayRequest for requesting Onboard endnode with gateway id.
+type OnboardEndnodeWithGatewayRequest struct {
```
**OnboardEndnodeWithGatewayRequest**
naming looks indecisive.
WithGateway doesn't adds much information to the struct but the makes the name redundant.

Suggested solutions:

#### Fully Qualified:

Prepare 2 structs for the different type of the content-type.
- OnboardEndnodeWithGatewayThingIDRequest
- OnboardEndnodeWithGatewayVendorThingIDRequest

#### Pack it in one struct.
- OnboardEndnodeRequest

The only difference between those content-types are gatewayVendorThingID and gatewayThingID.
Only accept either of them. (Somehow.)

yongpingchen replied:
Thanks @satoshikumano,  I prefer to use `OnboardEndnodeRequest`

satoshikumano commented on api.go:
```diff
@@ -189,6 +189,22 @@ type OnboardByOwnerRequest struct {
+}
+
+// OnboardEndnodeWithGatewayResponse for receiving response of onboarding endnode with gateway
+type OnboardEndnodeWithGatewayResponse struct {
```
Response is common.
**OnboardEndnodeResponse** would be fine.

satoshikumano commented:
@yongpingchen  please confirm.

yongpingchen committed:
rename struct

OnboardEndnodeWithGatewayRequest > OnboardEndnodeRequest
OnboardEndnodeWithGatewayResponse > OnboardEndnodeResponse

yongpingchen committed:
full qualify OnboardEndnodeWithGateway request

OnboardEndnodeWithGatewayThingIDRequest
OnboardEndnodeWithGatewayVendorThingIDRequest

yongpingchen commented:
@satoshikumano could you confirm?

satoshikumano commented on api.go:
```diff
@@ -189,6 +189,34 @@ type OnboardByOwnerRequest struct {
 
+// OnboardEndnodeBasicRequest is the command fields for
+// OnboardEndnodeWithGatewayThingIDRequest and OnboardEndnodeWithGatewayVendorThingIDRequest
+type OnboardEndnodeBasicRequest struct {
```
OnboardEndnodeRequestCommon

With 'Basic' it looks like GatewayThingID or GatewayVendorThingID is optional.

yongpingchen committed:
rename struct

OnboardEndnodeBasicRequest > OnboardEndnodeRequestCommon

yongpingchen commented:
thanks @satoshikumano I renamed it

satoshikumano commented on author.go:
```diff
@@ -229,7 +231,40 @@ func (a *APIAuthor) OnboardThingByOwner(request OnboardByOwnerRequest) (*Onboard
+
+// OnboardEndnodeWithGateway onboards an endnode
+// request must be either OnboardEndnodeWithGatewayVendorThingIDRequest or OnboardEndnodeWithGatewayThingIDRequest
+func (a *APIAuthor) OnboardEndnodeWithGateway(request interface{}) (*OnboardEndnodeResponse, error) {
```
OnboardEndnodeWithGatewayVendorThingID
OnboardEndnodeWithGatewayThingID

yongpingchen committed:
Add 2 new methods: OnboardEndnodeWithGatewayVendorThingID and OnboardEndnodeWithGatewayThingID, based on old method
